skfp/utils/splitters.py

```diff
@@ -0,0 +1,630 @@
+    This method, suggested in the MoleculeNet [2]_ paper, helps to test the model's ability to
+    generalize to entirely new scaffolds. MoleculeNet introduced the scaffold split as an approximation
+    to the time split, assuming that new molecules (test set) will be structurally different in terms of
+    scaffolds from the training set.
+
+    The split is fully deterministic, with the smallest scaffold sets assigned to the test
+    subset and the rest to the training subset.
+
+    The split fractions (train_size, test_size) must sum to 1.
+
+    Parameters
+    ----------
+    data : sequence
+        Sequence representing either SMILES strings or RDKit `Mol` objects.
+
+    additional_data: list[sequence]
+        Additional sequences to be split alongside the main data (e.g., labels or feature vectors).
+
+    train_size : float, default=None
+        The fraction of data to be used for the train subset. If None, it is set to 1 - test_size.
+
+    test_size : float, default=None
+        The fraction of data to be used for the test subset. If None, it is set to 1 - train_size.
+
+    include_chirality: bool, default=False
+        Whether to take chirality of molecules into consideration.
+
+    return_indices : bool, default=False
+        Whether the method should return the input object subsets, i.e. SMILES strings
+        or RDKit `Mol` objects, or only the indices of the subsets.
+
+    Returns
+    ----------
+    subsets : tuple[list, list]
+        A tuple of train and test subsets. The format depends on `return_indices`:
+        - if `return_indices` is False, returns lists of SMILES strings
+        or RDKit `Mol` objects depending on the input
+        - if `return_indices` is True, returns lists of indices
+    additional_data_splitted: list[sequence]:
+        the method may return any additional data splitted in the same way as
+        provided SMILES
+
+    References
+    ----------
+    .. [1] `Bemis, G. W., & Murcko, M. A.
+        "The properties of known drugs. 1. Molecular frameworks."
+        Journal of Medicinal Chemistry, 39(15), 2887-2893.
+        https://www.researchgate.net/publication/14493474_The_Properties_of_Known_Drugs_1_Molecular_Frameworks`_
+
+    .. [2] `Z. Wu, B. Ramsundar, E. N. Feinberg, J. Gomes, C. Geniesse, A. S. Pappu, K. Leswing, V. Pande
+        "MoleculeNet: A Benchmark for Molecular Machine Learning."
+        Chemical Science, 9(2), 513-530.
+        https://www.researchgate.net/publication/314182452_MoleculeNet_A_Benchmark_for_Molecular_Machine_Learning`_
+
+    """
+    _validate_split_sizes(train_size=train_size, test_size=test_size)
+    train_size, test_size = _fill_missing_sizes(train_size, test_size)
+    scaffolds = _create_scaffolds(list(data), include_chirality)
+    scaffold_sets = sorted(scaffolds.values(), key=len)
+
+    train_ids: list[int] = []
+    test_ids: list[int] = []
+
+    train_ids, test_ids = _split_ids(scaffold_sets, len(data), test_size)
```

Comment:
I would definitely prefer to have this here. If this causes cyclomatic complexity hook to fail, raise the file-wide limit in pre-commit hook config